libpod/in_memory_state.go

```diff
@@ -289,6 +293,86 @@ func (s *InMemoryState) RemovePod(pod *Pod) error {
 	return nil
 }
 
+// UpdatePod updates a pod's state from the backing database
+// As in-memory states have no database this is a no-op
+func (s *InMemoryState) UpdatePod(pod *Pod) error {
+	return nil
+}
+
+// AddContainerToPod adds a container to the given pod, also adding it to the
+// state
+func (s *InMemoryState) AddContainerToPod(pod *Pod, ctr *Container) error {
+	if !pod.valid {
+		return errors.Wrapf(ErrPodRemoved, "pod %s is not valid and cannot be added to", pod.ID())
+	} else if !ctr.valid {
+		return errors.Wrapf(ErrCtrRemoved, "container %s is not valid and cannot be added to the pod", ctr.ID())
+	}
+
+	if ctr.config.Pod != pod.ID() {
+		return errors.Wrapf(ErrInvalidArg, "container %s is not in pod %s", ctr.ID(), pod.ID())
+	}
+
+	// Add container to pod
+	if err := pod.addContainer(ctr); err != nil {
+		return err
+	}
+
+	// Add container to state
+	_, ok := s.containers[ctr.ID()]
+	if ok {
+		return errors.Wrapf(ErrCtrExists, "container with ID %s already exists in state", ctr.ID())
```

Comment:
Can we add the state value to the error?

Reply:
It's probably going to be the same container - duplicate IDs are incredibly, impossibly unlikely so this usually means you tried to add a container that was already added. I'll make the error message more clear about this.